File: python/mirage/accessor/nextcloud.py
import opendal

from mirage.accessor.base import Accessor
from mirage.resource.secrets import reveal_secret


class NextcloudAccessor(Accessor):

    def __init__(self, config) -> None:
        self.config = config

    def operator(self):
        config = self.config
        kwargs = {"endpoint": config.url}
        username = reveal_secret(config.username)
        if username:
            kwargs["username"] = username
        password = reveal_secret(config.password)
        if password:
            kwargs["password"] = password
        return opendal.AsyncOperator("webdav", **kwargs)

File: typescript/packages/node/src/accessor/nextcloud.ts
import type { Operator } from 'opendal'
import { Accessor } from '@struktoai/mirage-core/accessor/index'
import { VFSName } from '@struktoai/mirage-core/types'
import { loadOptionalPeer } from '../optional_peer.ts'
import type { NextcloudConfig } from '../vfs/nextcloud/config.ts'

export class NextcloudAccessor extends Accessor {
  readonly vfsName = VFSName.NEXTCLOUD
  private operatorPromise: Promise<Operator> | null = null

  constructor(readonly config: NextcloudConfig) {
    super()
  }

  operator(): Promise<Operator> {
    this.operatorPromise ??= this.createOperator()
    return this.operatorPromise
  }

  private async createOperator(): Promise<Operator> {
    const mod = await loadOptionalPeer(
      () => import('opendal') as Promise<{ Operator: typeof Operator }>,
      { feature: 'Nextcloud VFS', packageName: 'opendal' },
    )
    const options: Record<string, string> = { endpoint: this.config.url }
    if (this.config.username !== undefined) options.username = this.config.username
    if (this.config.password !== undefined) options.password = this.config.password
    return new mod.Operator('webdav', options)
  }
}
